Flip-Chip stack

sandokokosen opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->


### What are the issues this pull addresses (issue numbers / links)?
#491 

### Did you add tests to cover your changes (yes/no)?
No

### Did you update the documentation accordingly (yes/no)?
Yes (in the docstring)

### Did you read the CONTRIBUTING document (yes/no)?
Yes

### Summary
Enable the rendering of a two-chip stack and objects at a non-zero center_z in ANSYS.


### Details and comments
See discussion #491

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-metal?pullRequest=677) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-metal?pullRequest=677) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: warrench<br/>:x: Sandoko Kosen<br/><hr/>**Sandoko Kosen** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-metal?pullRequest=677) it.</sub>

ThomasGM4 commented:
Currently testing the code.

zlatko-minev commented:
Awesome work, @sandokokosen , @warrench , and team! Looking this over now

zlatko-minev commented on qiskit_metal/designs/design_flipchip.py:
```diff
@@ -17,14 +17,15 @@
 from .design_base import QDesign, Dict
 
-__all__ = ['DesignFlipChip']
+__all__ = ["DesignFlipChip"]
```
Very nice edits :)

sandokokosen commented:
Hi, really sorry for the long delay! I have since applied the yapf formatting. Hopefully there is no such issue anymore this time.

ThomasGM4 commented:
No apology necessary! It is great that you are contributing this code.

I will try and do some testing of this in the next few days. 
Was thinking, if you have the time, maybe a tutorial notebook showing how the new feature works would be helpful for users?

sandokokosen commented:
Hi Thomas,

That would indeed be helpful, unfortunately I won't be able to do it straightaway. Any example of tutorial notebook you have in mind?

ThomasGM4 commented:
https://github.com/Qiskit/qiskit-metal/blob/main/tutorials/4%20Analysis/A.%20Core%20-%20EM%20and%20quantization/4.02%20Eigenmode%20and%20EPR.ipynb

Could aim for this one but with a flip chip basically? *Think* the EPR should work, but will be an interesting test

sandokokosen commented:
Looks like an interesting notebook, I shall have a look at it in the next few weeks after I'm back from my travel. Might be a good occasion to try pyEPR too :)

sandokokosen commented:
Hello All,

Little update from my side: I'll start working on the notebook (and read more about Zlatko's EPR technique) this month or next :). I might have to add few more qcomponents along the way too for the notebook.

zlatko-minev commented:
Awesome, really great to hear. Feel free to post updates / questions here

sandokokosen commented:
Hi, I added a short tutorial notebook for the flipchip design class. I had to make an adjustment to the ansys_renderer render_chip function. Under the current configuration, if there is no component specified for that chip, ansys_renderer does not render that chip, which is a problem for the flipchip case because we always need both chips to be there. So I added few lines of codes to solve this. But this certainly looks like a quick fix rather than a good long-term solution. 

Another alternative involves giving the user more freedom to render all chips independent of whether there is a component on that particular chip that we want to render in ANSYS. This looks like it needs more modification, but I'm not sure if this is something that you guys want, so I went for the first option.

ThomasGM4 commented on qiskit_metal/qlibrary/resonator/readoutres_fc.py:
```diff
@@ -0,0 +1,189 @@
+
+    Inherits BaseQubit class
+
+    Readout resonator for the flipchip dev. Written for the flipchip tutorial.
```
Will want to expand the documentation here to mimic what is seen with other components.

Also thinking maybe not wanting BaseQubit as the parent class, as that makes renderers expect there to be a JosephsonJunction to be present in the component.

sandokokosen replied:
1) Thanks for the feedback thomas, sorry for the lack of explanatory docstrings as I thought that objects like this are not something you would like to keep in the long term since it lacks some generality that one would typically want to find when rendering a transmission line resonator. I can certainly improve the documentation in upcoming updates.
Actually, I wanted to draw the transmission line part of the resonator using QRoute and the circular head shape using the shapely function directly, then combine it within one class object. But at the time, I wasn't super clear how things would be grouped and named in ansys, and without taking the time to test through things and browse through QRoute implementations, I reverted to a more simpler way, which was to draw everything using the shapely function directly, something I understood a bit better. It is likely though that I'll phase out this way of drawing cpw resonators in the future in favor of using QRoute.

2) I used BaseQubit as the parent class and set the _qgeometry_table_junction in the component_metadata dictionary as False, which I thought should let the renderers know that there are no JosephsonJunction in the object? I wasn't exactly sure why I used BaseQubit as the parent class. It is very likely that I was thinking of rendering a tunable cpw resonator (with a jj). In any case, I can instead use QComponent as the parent class, and I think there should be no problem. Will change this in upcoming updates.

3) I wasn't exactly sure where I should place this resonator object in the qlibrary folder. For the moment I've intentionally created a specific folder for this, but if you feel that it should belong somewhere else, please let me know. The same for the tutorial notebook, I wasn't sure which folder it should belong to.

priti-ashvin-shah-ibm commented:
@sandokokosen  - Thank you for taking time to contribute to Qiskit Metal.  

I am looking at the CI testing.  I noticed the image below.

There is a test to check the number of default options on line 485 in https://github.com/Qiskit/qiskit-metal/blob/main/qiskit_metal/tests/test_qlibrary_2_options.py.  You have added a default option of chip="main".  So the test should  now expect 5 options, instead of 4. 

To get test to pass, please update the test on line 485 to be 5 instead of 4. 

https://github.com/Qiskit/qiskit-metal/blob/main/qiskit_metal/tests/test_qlibrary_2_options.py


![image](https://user-images.githubusercontent.com/74020801/151035110-8da53cd9-5125-4326-ac79-35108abbc92e.png)

priti-ashvin-shah-ibm commented:
> [![CLA assistant check](https://camo.githubusercontent.com/4c970efeed75dd01f6778bf7e33dec85e4171375cb69acb0927bfc9d437d78d9/68747470733a2f2f636c612d617373697374616e742e696f2f70756c6c2f62616467652f6e6f745f7369676e6564)](https://cla-assistant.io/Qiskit/qiskit-metal?pullRequest=677) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-metal?pullRequest=677) before we can accept your contribution.**2** out of **4** committers have signed the CLA.✅ zlatko-minev✅ warrench❌ sandokokosen❌ ThomasGM4You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-metal?pullRequest=677) it.

Hi @sandokokosen , we have had an issue with the CLA, would you please sign it again?

sandokokosen commented:
Hi Priti, thank you so much for figuring out the issue with the CI testing, I will get it fixed. Will fix the linting issue, and also try to sign the CLA again.

priti-ashvin-shah-ibm commented:
> Hi Priti, thank you so much for figuring out the issue with the CI testing, I will get it fixed. Will fix the linting issue, and also try to sign the CLA again.

My pleasure.  We are trying to make it easier for people to contribute code.  
There was an unexpected issue with the CLA bot.  It has been fixed but unfortunately, some people will have to resign the CLA.

priti-ashvin-shah-ibm commented:
> Hi Priti, thank you so much for figuring out the issue with the CI testing, I will get it fixed. Will fix the linting issue, and also try to sign the CLA again.

Regarding the linting/formatting , this may help from link  below.  
 
![image](https://user-images.githubusercontent.com/74020801/151250075-2efa7089-b21f-45b0-a095-df559ab66b78.png)

https://github.com/Qiskit/qiskit-metal/blob/main/CONTRIBUTING.md

ThomasGM4 reviewed (approved):
Just finished testing old designs and your notebook. All looks good and runs well. Only error is the little one with EPR which is nothing to do with your PR.

I think ready to merge.
